private/rules/artifact.bzl

```diff
@@ -1,19 +1,22 @@
 load("//:specs.bzl", "parse")
 load("//private:constants.bzl", "DEFAULT_REPOSITORY_NAME")
+load("//private:coursier_utilities.bzl", "get_simple_coord")
 
 def artifact(a, repository_name = DEFAULT_REPOSITORY_NAME):
-    artifact_obj = _parse_artifact_str(a) if type(a) == "string" else a
-    return "@%s//:%s" % (repository_name, _escape(artifact_obj["group"] + ":" + artifact_obj["artifact"]))
+    artifact_str = _make_artifact_str(a) if type(a) != "string" else _ensure_version(a)
+    return "@%s//:%s" % (repository_name, _escape(get_simple_coord(artifact_str)))
 
 def maven_artifact(a):
     return artifact(a, repository_name = DEFAULT_REPOSITORY_NAME)
 
 def _escape(string):
     return string.replace(".", "_").replace("-", "_").replace(":", "_")
 
-def _parse_artifact_str(artifact_str):
-    pieces = artifact_str.split(":")
-    if len(pieces) == 2:
-        return {"group": pieces[0], "artifact": pieces[1]}
-    else:
-        return parse.parse_maven_coordinate(artifact_str)
+def _ensure_version(a):
+    # fake :version part is due to get_simple_coord not supporting unversioned coords
+    return a if len(a.split(":")) > 2 else (a + ":fake_version")
+
+def _make_artifact_str(artifact_obj):
+    # TODO: add support for optional type, classifier and version parts
```

Comment:
gets a little complicated with dictionary case as there are multi-level optional fields so if `packaging` is absent but `classifier` is present would need to apply some default value to packaging, this is a new use-case and can be added later, for now dict version still only handles `group + artifact` only

Reply:
calling `fail()` in cases where `packaging` is absent but `classifier` is present could be an option too